peer/example_test.go

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2015 The btcsuite developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package peer_test
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/btcsuite/btcd/addrmgr"
+	"github.com/btcsuite/btcd/peer"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// zeroHash is the zero value hash (all zeros).  It is defined as a convenience.
+var zeroHash wire.ShaHash
+
+// Not implemented.
+func lookupFunc(host string) ([]net.IP, error) {
+	return nil, errors.New("not implemented")
+}
+
+// newestSha returns the latest known block to this peer.  It returns a
+// hard-coded hash and height since this is only used for example code.
+func newestSha() (*wire.ShaHash, int32, error) {
+	hashStr := "14a0810ac680a3eb3f82edc878cea25ec41d6b790744e5daeef"
+	hash, err := wire.NewShaHashFromStr(hashStr)
+	if err != nil {
+		return nil, 0, err
+	}
+	return hash, 234439, nil
+}
+
+// This example demonstrates initializing both inbound and outbound peers.  An
+// inbound peer listening on mainnet port i.e 8333 is started first, then an
+// outbound peer is connected to it.
+// Peers negotiate protocol by exchanging version and verack messages.  For
+// demonstration, a simple handler for version message is attached to both
+// peers.
+func Example_peerConnection() {
+	addrMgr := addrmgr.New("test", lookupFunc)
+	peerCfg := &peer.Config{
+		UserAgentName:    "peer",
+		UserAgentVersion: "1.0",
+	}
+	listening := make(chan struct{})
+	go func() {
+		l1, err := net.Listen("tcp", "127.0.0.1:8333")
+		if err != nil {
+			fmt.Printf("Listen: error %v\n", err)
+			return
+		}
+		listening <- struct{}{}
+		c1, err := l1.Accept()
+		if err != nil {
+			fmt.Printf("Accept: error %v\n", err)
+			return
+		}
+		p1 := peer.NewInboundPeer(newestSha, 0, wire.MainNet, wire.SFNodeNetwork, addrMgr, peerCfg, c1)
+		p1.AddVersionMsgListener("handleVersionMsg", func(p *peer.Peer, msg *wire.MsgVersion) {
+			fmt.Println("inbound: received version")
+		})
+		err = p1.Start()
+		if err != nil {
+			fmt.Printf("Start: error %v\n", err)
+			return
+		}
+	}()
+	<-listening
```

Comment:
This will block forever if there is an error establishing the listener in the goroutine (which could easily happen if btcd it running already listening on 8333, for example).